engine/src/main/java/org/hibernate/search/v6poc/engine/impl/SearchMappingRepositoryBuilderImpl.java

```diff
@@ -142,34 +146,62 @@ public SearchMappingRepository getBuiltResult() {
 	private static class MapperContribution<C, M extends MappingImplementor> {
 
 		private final MapperFactory<C, M> mapperFactory;
+		private final Set<MappableTypeModel> typesLeftToProcess = new LinkedHashSet<>();
+		private final Set<MappableTypeModel> typesBeingProcessed = new LinkedHashSet<>();
+		private final Set<MappableTypeModel> freezedTypes = new HashSet<>();
 		private final Map<MappableTypeModel, TypeMappingContribution<C>> contributionByType = new HashMap<>();
 
 		MapperContribution(MapperFactory<C, M> mapperFactory) {
 			this.mapperFactory = mapperFactory;
 		}
 
 		public void update(MappableTypeModel typeModel, String indexName, C contributor) {
+			if ( freezedTypes.contains( typeModel ) ) {
+				throw new AssertionFailure(
+						"Attempt to add a mapping contribution for a type which has already been mapped."
+						+ " There is a bug in the mapper implementation."
```

Comment:
Might be a good idea to add some info about the typeModel to the exception.

Reply:
Right, done.